packages/framework-provider-aws/src/library/searcher-adapter.ts

```diff
@@ -32,62 +31,150 @@ export async function searchReadModel(
   return result.Items ?? []
 }
 
-function buildFilterExpression(filters: Record<string, Filter<any>>): string {
+function buildExpression(filters: FilterFor<any>): any {
+  const expresion = buildFilterExpression(filters)
+  const names = buildExpressionAttributeNames(filters)
+  resetAttrCount()
+  const values = buildExpressionAttributeValues(filters)
+  resetAttrCount()
```

Comment:
I see you call `resetAttrCount()` after `buildExpressionAttributeNames`, but that function doesn't use the global array that is reset. I guess you wanted to call it after `buildFilterExpression`.

Anyway, I think that using a global array (to this file) and knowing when I should reset it (it is not clear) is dangerous and can lead to bugs. Let me make you a proposal:

Instead of having a global array and resetting it after calling each of these two functions, _Why not creating two different empty arrays and passing them as parameters (this wouldn't be weird as the functions are recursive, so it is a common pattern to pass initial values as parameters)?_

The code would end up like this:
```suggestion
  const expresion = buildFilterExpression(filters, [])
  const names = buildExpressionAttributeNames(filters)
  const values = buildExpressionAttributeValues(filters, [])
```

This would erase possible subtle bugs when modifying this code and misplacing those "reset" calls or forget to add them, as it is not clear why we need to call them right here.

Check the other comments where I give hints of the other needed changes for this proposal

Reply:
This is awesome @alvaroloes! 💯  I just implemented it and even improve it a little, I'll show you shortly!